### PR TITLE
chore ::> method on the BroadcastInvokeTxnType interface to constrain it.

### LINKED ISSUE #569: Constrain BroadcastInvokeTxnType
Define a method on the BroadcastInvokeTxnType interface and implement if for the three different BroadcastInvokevXTnx types. This improves UX so that users won't be able to pass anything in like they are now. Bonus points for simplicity!

### PR DESCRIPTION
**Pull request by PsychoPunkSage**
closes #569
added `InvokeType() string` function in `BroadcastInvokeTxnType` interface.
and extrapolated it for `BroadcastInvokev0Txn`, `BroadcastInvokev1Txn` and `BroadcastInvokev3Txn`

### CODE REVIEW

**Comment by PsychoPunkSage** [comment]
Hi @rianhughes 
Am I doing what was asked in issue https://github.com/NethermindEth/starknet.go/issues/569 ?
(I have to update other files which previously considered `BroadcastInvokeTxnType` as an empty interface.... I will do it once I get to know whether I'm doing things correct way.. )

File: rpc/types_transaction.go [hunk 1]
```diff
@@ -43,6 +43,10 @@ type InvokeTxnV0 struct {
 	FunctionCall
 }
 
+func (tx InvokeTxnV0) InvokeType() string {
+	return "InvokeTxn: V0"
+}
+
```
Comment (rianhughes) [inline rpc/types_transaction.go]: We don't need to define this method on the InvokeTxnVX types.

If you are getting errors when trying to pass in (eg) InvokeV1 txns in TestAddInvokeTransaction, then you just need to wrap the InvokeTxnsV1 in BroadcastInvokeV1Txn
`InvokeTx:     BroadcastInvokev1Txn{InvokeTxnV1{SenderAddress: new(felt.Felt).SetUint64(123)}},`

File: rpc/types_broadcast_transaction.go [hunk 2]
```diff
@@ -40,14 +42,26 @@ type BroadcastInvokev0Txn struct {
 	InvokeTxnV0
 }
 
+func (tx BroadcastInvokev0Txn) InvokeType() string {
```
Comment (rianhughes) [inline rpc/types_broadcast_transaction.go]: Can you also do this for the broadcastDeclare and broadcastDeployAccount types please?

Comment (PsychoPunkSage) [reply rpc/types_broadcast_transaction.go]: Sure I will do that
Should I create a separate PR for them?? To keep things segregated?

File: rpc/types_transaction.go (hunk 1 above)
Comment (PsychoPunkSage) [reply rpc/types_transaction.go]: Ok Got it...
I will make changes and push it again...

Comment (PsychoPunkSage) [reply rpc/types_transaction.go]: Resolved it....
You can have a look at it...

Comment (PsychoPunkSage) [reply rpc/types_transaction.go]: Hi @rianhughes 
I did what you told.... But there is a weird error which I'm facing in `rpc/write_test.go` >> `TestAddInvokeTransaction`

**For Mock:**
1. `InvokeTx: BroadcastInvokev3Txn{......`
I got this error 
```
=== RUN   TestAddInvokeTransaction
--- FAIL: TestAddInvokeTransaction (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x88dc32]
```

2. `InvokeTx:  BroadcastInvokev1Txn{InvokeTxnV1{}},`
I got this error 
```
=== RUN   TestAddInvokeTransaction
--- FAIL: TestAddInvokeTransaction (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x88d5f8]
```

3. `InvokeTx:     BroadcastInvokev1Txn{InvokeTxnV1{SenderAddress: ......}}`
I got this error 
```
=== RUN   TestAddInvokeTransaction
    /home/psychopunk_sage/dev/Workplace/Nethermind/starknet.go/rpc/write_test.go:174:
                Error Trace:    /home/psychopunk_sage/dev/Workplace/Nethermind/starknet.go/rpc/write_test.go:174
                Error:          Not equal:
                                expected: &rpc.RPCError{Code:63, Message:"An unexpected error occurred", Data:"Something crazy happened"}
                                actual  : &rpc.RPCError{Code:-32603, Message:"Internal Error", Data:"0  <nil>\n"}
```

Comment (PsychoPunkSage) [reply rpc/types_transaction.go]: Like all these errors are unusual... they shouldn't occur, don't know what is the issue (I'm trying to find the solution.)

File: rpc/types_broadcast_transaction.go (hunk 2 above)
Comment (rianhughes) [reply rpc/types_broadcast_transaction.go]: Might as well do them all at once  if you don't mind

File: rpc/types_transaction.go (hunk 1 above)
Comment (rianhughes) [reply rpc/types_transaction.go]: Let me know if you get these errors after trying the new approach as outlined by @thiagodeev in the internal discussions

Comment (rianhughes) [reply rpc/types_transaction.go]: The test calls `mock_starknet_addInvokeTransaction` which currently expects eg `InvokeTxnV3`, but we are passing in `BroadcastDeclareTxnV3`. You just need to update the switch to use the correct types

Comment (PsychoPunkSage) [reply rpc/types_transaction.go]: Thanks Its working now....

File: rpc/types_broadcast_transaction.go (hunk 2 above)
Comment (PsychoPunkSage) [reply rpc/types_broadcast_transaction.go]: Mmm
I will create a separate one to keep things clean from my side :)

**Comment by PsychoPunkSage** [comment]
hi @rianhughes @thiagodeev 
Please have a look at the changes... I suppose this is what you are looking for.

**Review by rianhughes — APPROVED** [review]

File: rpc/types_broadcast_transaction.go [hunk 3]
```diff
@@ -40,14 +42,26 @@ type BroadcastInvokev0Txn struct {
 	InvokeTxnV0
 }
 
+func (tx BroadcastInvokev0Txn) GetCalldata() []*felt.Felt {
+	return tx.InvokeTxnV0.FunctionCall.Calldata
```
Comment (thiagodeev) [inline rpc/types_broadcast_transaction.go]: NIT: you can also set this to `tx.Calldata` thanks to [Go Struct Embedding](https://gobyexample.com/struct-embedding)

Comment (PsychoPunkSage) [reply rpc/types_broadcast_transaction.go]: Done...

**Pull request merged**